### DISCUSSION TITLE
How to set a 'condition' in a `model` like WebPPL?

### DISCUSSION THREAD

**Discussion by Flawless1202** [ReactiveBayes/RxInfer.jl#332] [Q&A]
I have read the tutorial called [Probabilistic Models of Cognition](https://probmods.org/). And I have read an example in Section [Causal and statistical dependence](https://probmods.org/chapters/dependence.html) as follows:
```JavaScript
var marg = Infer({method: 'enumerate'}, function() {
  var smokes = flip(0.2)
  var lungDisease = (smokes && flip(0.1)) || flip(0.001)
  var cold = flip(0.02)
  var cough = (cold && flip(0.5)) || (lungDisease && flip(0.5)) || flip(0.001)
  var fever = (cold && flip(0.3)) || flip(0.01)
  var chestPain = (lungDisease && flip(0.2)) || flip(0.01)
  var shortnessOfBreath = (lungDisease && flip(0.2)) || flip(0.01)

  condition(cough)
  return {cold: cold, lungDisease: lungDisease}
})

viz.marginals(marg)
```
I try to realize the `condition` function in `RxInfer.jl`. I follow the example in this [tutorial](https://reactivebayes.github.io/RxInfer.jl/stable/examples/advanced_examples/Assessing%20People%20Skills/). Here is my code:
```Julia
using RxInfer, Distributions

struct Cond end
@node Cond Stochastic [out, in]
@rule Cond(:in, Marginalisation) (q_out::PointMass,) = begin
    return Bernoulli(mean(q_out))
end

@model function f(cond)
    smokes ~ Bernoulli(0.2)
    lung_disease ~ (smokes && Bernoulli(0.1)) || Bernoulli(0.001)
    cold ~ Bernoulli(0.02)
    cough ~ (cold && Bernoulli(0.5)) || (lung_disease && Bernoulli(0.5)) || Bernoulli(0.001)
    cond .~ Cond(cough)
end

result = infer(model = f(), data = (cond = [true],))
```
And I get the true answer. But in my implementation, the input of `Cond` must be a distribution. When I use a Bool expression such as  `cond .~ Cond(cough == true)`, I get the error as follows:
```Bash
Delta node `==` requires meta specification with the `where { meta = ... }` in the `@model` macro or with the separate `@meta` specification. See documentation for the `DeltaMeta`.
```

Is there any way to fix this error and make the `Cond` function work well? @ThijsvdLaar  @bvdmitri Thanks for your reply!

**Comment by albertpod** [comment]
Hi @Flawless1202! Thanks for trying out RxInfer.jl!

I’m not entirely sure I understand your question. The variable “cough” is a random variable drawn from the lengthy expression. You then pass this random variable to a `Cond` node (not a function in Julia sense). Could you elaborate a bit more on what you’re trying to achieve?

**Reply by albertpod** [reply]
If I understand you correctly, the purpose of the inference in this model is to determine the probabilities of the underlying factors (in this case, having a cold and having lung disease) given the observation that the person has a cough. I think you almost got it right, but perhaps the model should look this way?

```julia
@model function f(cough)
    smokes ~ Bernoulli(0.2)
    lung_disease ~ (smokes && Bernoulli(0.1)) || Bernoulli(0.001)
    cold ~ Bernoulli(0.02)
    cond ~ (cold && Bernoulli(0.5)) || (lung_disease && Bernoulli(0.5)) || Bernoulli(0.001)
    cough .~ Cond(cond)
end

result = infer(model = f(), data = (cough = [true],))
```

So now, your observation is a cough, from which you can compute the posterior over the factors.

**Reply by Flawless1202** [reply]
Thanks for your reply. This example may be a little bit misleading. I will give one more example from [Probabilistic Models of Cognition](https://probmods.org/) (Section [Conditioning](https://probmods.org/chapters/conditioning.html)):
```JavaScript
var model = function () {
  var A = flip()
  var B = flip()
  var C = flip()
  var D = A + B + C
  //add the desired assumption:
  condition(D >= 2)
  return {'A': A}
};
var dist = Infer(model)
```
Here ` condition(D >= 2)` sets the observation `D` always greater than or equal to 2. Is there any way to realize the `condtion(D >= 2)` in RxInfer.jl?

**Reply by albertpod** [reply]
@Nimrais can we bypass Bernoulli RVs to delta node and make use of new projection method?

**Reply by Nimrais** [reply]
yes, we can, but I am not sure that I understood what model is doing. The links does not open for me.

Can @Flawless1202 you drop graphical representation of the model here?

**Reply by Nimrais** [reply]
mainly what are we observing and what are we inferring?

**Reply by wouterwln** [reply]
It looks to me that we need the `Transition` node for this problem. I'm interpreting `(smokes && Bernoulli(0.1)) || Bernoulli(0.001)` as "If you smoke, there's a 10% chance on lung disease, otherwise there is 0.1% chance" You'd realize that with
```julia
@model function f(cough)
    smokes ~ Bernoulli(0.2)
    lung_disease ~ Transition(smokes, [0.1 0.001; 0.9 0.999])
    cold ~ Bernoulli(0.2)
    ...
end
``` 
I've deliberately put 3 dots for the rest of the model, because the expression for `cond` is a Contingency table between 3 Bernoulli random variables. We can implement this with the `TransitionMixture` node in (https://github.com/ReactiveBayes/ReactiveMP.jl/pull/407) this PR, but this example warrants from a general Contingency node that takes `N` nonparametric discrete distributions.

**Reply by albertpod** [reply]
@Flawless1202, essentially what you’re saying is that you want to sum three Bernoulli random variables, then apply a condition (sum > 2), and observe the outcome. Therefore, your inference problem becomes: given the outcome, what is the posterior distribution over those three Bernoulli random variables?

**Reply by Nimrais** [reply]
@Flawless1202 Regarding your question about RxInfer implementation of this model

> 
```
var model = function () {
  var A = flip()
  var B = flip()
  var C = flip()
  var D = A + B + C
  //add the desired assumption:
  condition(D >= 2)
  return {'A': A}
};
var dist = Infer(model)
```

It can be almost one-to-one translated into RxInfer like this

```
using RxInfer, Distributions
using ExponentialFamilyProjection
using Manopt

struct Cond end

@node Cond Stochastic [out, in]

@rule Cond(:in, Marginalisation) (q_out::PointMass,) = begin
    return Bernoulli(mean(q_out))
end

function f(x, y, z)
    return x + y + z >= 2
end

@model function model(cond)
    A ~ Bernoulli(0.5)
    B ~ Bernoulli(0.5)
    C ~ Bernoulli(0.5)
    cond .~ Cond(f(A, B, C))
end

delta_meta = @meta begin
    f() -> CVIProjection()
end

result = infer(model = model(), data = (cond = [0.999],), meta = delta_meta)
```

If you're wondering why we need `@meta`, you can check how to implement inference around deterministic nodes [here](https://reactivebayes.github.io/RxInfer.jl/stable/manuals/inference/delta-node/). Generally, when you have a deterministic relationship in your model, you need to specify how to run inference around it. `CVIProjection()` is the most generic approach that will work in almost all scenarios.

The only tricky part here is setting `cond = [0.999]`. We have only one observation, and to avoid problems with hard constraints, I softened it slightly.
